File: lab/javascript/arrays_e_listas/sobreposicao.js
// Normaliza um intervalo [a,b] -> [min, max]
function norm([a, b]) {
  return a <= b ? [a, b] : [b, a];
}

// 1) Dois intervalos se sobrepõem? (conta "encostar" como sobreposição)
function intervalosSobrepostos(i1, i2) {
  const [a, b] = norm(i1);
  const [c, d] = norm(i2);
  return !(b < c || d < a);
}

// 2) Algum de l1 sobrepõe algum de l2?
function listasTemSobreposicao(l1, l2) {
  for (const i1 of l1) {
    for (const i2 of l2) {
      if (intervalosSobrepostos(i1, i2)) return true;
    }
  }
  return false;
}

// 3) Mesclar intervalos
function mesclarIntervalos(intervalos) {
  if (intervalos.length === 0) return [];
  // normaliza cada intervalo
  const normed = intervalos.map(norm);
  // ordena por início
  normed.sort((x, y) => x[0] - y[0]);

  const res = [normed[0].slice()];
  for (let i = 1; i < normed.length; i++) {
    const [ini, fim] = normed[i];
    const last = res[res.length - 1];
    if (ini <= last[1]) {
      // sobrepõe → estende o fim
      last[1] = Math.max(last[1], fim);
    } else {
      res.push([ini, fim]);
    }
  }
  return res;
}

// --- Exemplos rápidos ---
console.log(intervalosSobrepostos([1,5],[4,9]));   // true
console.log(intervalosSobrepostos([1,5],[6,10]));  // false
console.log(listasTemSobreposicao([[1,3],[10,20]], [[2,5],[30,40]])); // true
console.log(mesclarIntervalos([[1,3],[2,6],[8,10],[9,11]])); // [[1,6],[8,11]]

File: lab/c/arrays_e_listas/sobreposicao.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long long start;
    long long end;
} Interval;

// Normaliza um intervalo para start <= end
static inline Interval norm(Interval I) {
    if (I.start <= I.end) return I;
    Interval J = { I.end, I.start };
    return J;
}

// 1) Dois intervalos se sobrepõem? (conta "encostar" como sobreposição)
int intervalos_sobrepostos(Interval A, Interval B) {
    A = norm(A); B = norm(B);
    // sobrepõe se NÃO (A.end < B.start || B.end < A.start)
    return !(A.end < B.start || B.end < A.start);
}

// 2) Algum de l1 sobrepõe algum de l2?
int listas_tem_sobreposicao(const Interval *L1, int n1, const Interval *L2, int n2) {
    for (int i = 0; i < n1; i++) {
        for (int j = 0; j < n2; j++) {
            if (intervalos_sobrepostos(L1[i], L2[j])) return 1;
        }
    }
    return 0;
}

// Comparador para qsort (ordena por start)
int cmp_interval(const void *pa, const void *pb) {
    const Interval *a = (const Interval *)pa;
    const Interval *b = (const Interval *)pb;
    if (a->start < b->start) return -1;
    if (a->start > b->start) return 1;
    if (a->end   < b->end)   return -1;
    if (a->end   > b->end)   return 1;
    return 0;
}

// 3) Mesclar intervalos (merge)
// - intervals: entrada (será copiada/normalizada internamente)
// - n: tamanho da entrada
// - *out: ponteiro de saída alocado com os intervalos mesclados
// - retorna a quantidade de intervalos em *out
int mesclar_intervalos(const Interval *intervals, int n, Interval **out) {
    if (n <= 0) { *out = NULL; return 0; }

    // cópia + normalização
    Interval *v = (Interval *)malloc(sizeof(Interval) * n);
    if (!v) return -1;
    for (int i = 0; i < n; i++) v[i] = norm(intervals[i]);

    // ordena por start
    qsort(v, n, sizeof(Interval), cmp_interval);

    // vetor de saída (máx n)
    Interval *res = (Interval *)malloc(sizeof(Interval) * n);
    if (!res) { free(v); return -1; }

    int m = 0;
    res[m++] = v[0];

    for (int i = 1; i < n; i++) {
        Interval cur = v[i];
        Interval *last = &res[m - 1];

        if (cur.start <= last->end) {
            // sobrepõe → estende o fim
            if (cur.end > last->end) last->end = cur.end;
        } else {
            res[m++] = cur;
        }
    }

    free(v);
    *out = res;
    return m;
}

// --- Exemplo de uso ---
int main(void) {
    // 1) Dois intervalos
    Interval A = {1, 5}, B = {4, 9}, C = {6, 10};
    printf("A x B: %d\n", intervalos_sobrepostos(A, B)); // 1
    printf("A x C: %d\n", intervalos_sobrepostos(A, C)); // 0

    // 2) Listas
    Interval L1[] = {{1,3}, {10,20}};
    Interval L2[] = {{2,5}, {30,40}};
    printf("L1 x L2: %d\n", listas_tem_sobreposicao(L1, 2, L2, 2)); // 1

    // 3) Mesclar
    Interval X[] = {{1,3},{2,6},{8,10},{9,11}};
    Interval *Y = NULL;
    int ny = mesclar_intervalos(X, 4, &Y);
    if (ny >= 0) {
        printf("Mesclados (%d): ", ny);
        for (int i = 0; i < ny; i++) {
            printf("[%lld,%lld]%s", Y[i].start, Y[i].end, (i+1<ny)? " ":"\n");
        }
        free(Y);
    }

    return 0;
}
